src/route/dataset-route.ts

```diff
@@ -0,0 +1,176 @@
+            headers: undefined,
+            data: undefined,
+            errors: [
+                {
+                    field: 'csv',
+                    message: 'No CSV data available'
+                }
+            ]
+        });
+        return;
+    }
+    const internalName: string = req.body?.internal_name;
+    if (!internalName) {
+        res.status(400);
+        res.json({
+            success: false,
+            errors: [
+                {
+                    field: 'internal_name',
+                    message: 'No internal name for the dataset has been provided'
+                }
+            ]
+        });
+        return;
+    }
+    const dataset = Dataset.createDataset(internalName, 'BetaUser');
+    const saved_dataset_record = await dataset.save();
+    const uploadDTO = await uploadCSV(req.file?.buffer, saved_dataset_record);
+    if (!uploadDTO.success) {
+        res.status(400);
+    }
+    res.json(uploadDTO);
+});
+
+apiRoute.get('/', async (req, res) => {
+    const datasets = await Dataset.find();
+    const fileList: FileDescription[] = [];
+    for (const dataset of datasets) {
+        fileList.push({
+            internal_name: dataset.internalName,
+            id: dataset.id
+        });
+    }
+    res.json({ filelist: fileList });
+});
+
+apiRoute.get('/:dataset', async (req, res) => {
+    const datasetID = req.params.dataset;
+    if (datasetID === undefined || datasetID === null) {
+        res.status(404);
+        res.json({ message: 'Dataset not found... file is null or undefined' });
+        return;
+    }
+    const dataset = await Dataset.findOneBy({ id: datasetID });
+    if (!dataset) {
+        res.status(404);
+        res.json({ message: 'Dataset not found... dataset id is invalid' });
+        return;
+    }
+    const datafiles = await dataset.datafiles;
+    if (datafiles.length < 1) {
+        res.status(404);
+        res.json({ message: 'Dataset has no datafiles attached' });
+        return;
+    }
+    const dto = await datasetToDatasetDTO(dataset);
+    res.json(dto);
+});
+
+apiRoute.get('/:dataset/csv', async (req, res) => {
+    const dataLakeService = new DataLakeService();
+    const datasetID = req.params.dataset;
+    if (datasetID === undefined || datasetID === null) {
+        res.status(404);
+        res.json({ message: 'Dataset not found... You must specify a dataset ID' });
+        return;
+    }
+    const dataset = await Dataset.findOneBy({ id: datasetID });
+    if (dataset === undefined || dataset === null) {
+        res.status(404);
+        res.json({ message: 'Dataset not found... Dataset ID not found in Database' });
+        return;
+    }
+    const datafiles = await dataset.datafiles;
+    const fileToDownload: Datafile | undefined = datafiles
+        .sort(
+            (first: Datafile, second: Datafile) =>
+                new Date(second.creationDate).getTime() - new Date(first.creationDate).getTime()
+        )
+        .shift();
```

Comment:
Why is it only possible to get the most recently uploaded file back again?

Reply:
I've yet to see a use case where I need to a previous version and for the purposes of return a single object I just went with the latest version which preserving records of previous versions.   Against subject to change as the understanding of the user needs improves.